### PR TITLE
[WIP] Speed up using flattened tensors

### LINKED ISSUE #11053: [RFC] Speed up Torch AMP and cliip grad norm operations by flattening all the tensors behind the scene
## 🚀 Feature
- Flatten all the tensors for a given model (per parameter group for instance) in the background (not user visible)
- Replace the per parameter tensors by tensor views
- Speed up:
- - optimizer step (per parameter group, in most cases it will become just a few matmuls)
- - torch AMP grad scaler (one kernel on each way instead of one per parameter)
- - clip grad norm

edit: 
Depending on the targeted speed up, the above would need to apply to the parameters, the grads, or both

### Motivation
Per parameter gradient scaling (same scale for all) and grad norm take a very significant time, benchmarked to 15-20% on a single GPU with [this small test](https://github.com/facebookresearch/xformers/blob/main/examples/microGPT.py#L314). Doing the above would be the equivalent to fusing multiple kernel calls into one, which is provably faster

### Pitch
- make this an option for lightning (this will break if the user changes tensors manually during training, which is probably not a good idea anyway, but there are valid corner cases)
- use a [flat param wrapper](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/misc/flatten_params_wrapper.py#L39). This is done in several places in [Fairscale](https://github.com/facebookresearch/fairscale/), if that helps
- do the same with the grads (**this is definitely a limitation of the proposal, open for other suggestions, my memory is a bit fuzzy around that**. This nukes the grad=None option) 
- derive a custom grad scaler from pytorch's which works on the flat buffer instead of working on all the parameters
- change the anomaly detection to work on the flat buffer, instead of going through all the grads

### Alternatives
Not doing the above, that's fine but slow

cc @borda @akihironitta @SeanNaren @tchaton

### PR DESCRIPTION
**Pull request by rishabbala**
## What does this PR do?
Add a FlatParamsWrapper based on [Fairscale](https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/misc/flatten_params_wrapper.py#L39) to flatten parameters based on groups

Fixes #11053

<summary>TODO</summary>

- [ ] Modify FlatParamsWrapper
- [ ] Support flattening over multiple groups of parameters
- [ ] Verify flattened model reaches similar accuracy to unflattened version
- [ ] Test for multi-device training
- [ ] Saving and loading state_dict

<summary>Results</summary>

- [X] Verified flattening a small model on MNIST data, using multi-GPUs
- [X] Runtime:
                &emsp; Unflattened: 195s
                &emsp; Flattened: 187s

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

Current code is a draft. Just want to confirm if this is what we want, and if the PR is in the right direction
@Borda @tchaton @carmocca

### CODE REVIEW

**Comment by rishabbala** [comment]
Hi @tchaton @Borda @carmocca, I've updated the code for using Flat Tensors. Here is an example snippet of how to use the FlatParamsWrapper. This would make sure that each parameter group provided is flattened into a single tensor. I still have to test it completely, but would be open to comments or suggestions before that.

```
import torch
import torch, torch.nn as nn, torch.utils.data as data, torchvision as tv, torch.nn.functional as F
import lightning as L
import time

start = time.time()
class Model(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.model = nn.Sequential(
            nn.Linear(28 * 28, 500),
            nn.ReLU(),
            nn.Linear(500, 100),
            nn.ReLU()
        )
        self.lin = nn.Linear(100, 10)

    def forward(self, x):
        x = self.model(x.view(x.shape[0], -1))
        x = self.lin(x)
        return x 

    def training_step(self, batch, batch_idx):
        x, y = batch
        x_hat = self(x)
        loss = F.cross_entropy(x_hat, y)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam([
            {'params': self.model.parameters(), 'lr': 1e-3},
            {'params': self.lin.parameters(), 'lr': 1e-5}
            ])
        return optimizer


model = Model()
flat_model = L.FlattenParamsWrapper(module=model, param_list=[list(model.model.parameters()), list(model.lin.parameters())])

dataset = tv.datasets.MNIST(".", download=True, transform=tv.transforms.ToTensor())
train, val = data.random_split(dataset, [55000, 5000])

trainer = L.Trainer(max_epochs=2)
trainer.fit(flat_model, data.DataLoader(train), data.DataLoader(val))
```

**Comment by carmocca** [comment]
A complex feature like this would need to be motivated by strong benchmarks that back it up. Particularly with large models such as those supported in https://github.com/Lightning-AI/lit-gpt. I see the small MNIST benchmark at the top but that wouldn't be enough. Do you have access to hardware that could carry out these experiments? 

I wonder if this is relevant after the addition of `foreach=` flag in optimizers with the most recent Torch versions (where the number of kernel launches is reduced at the cost of higher memory usage).

It's also not clear how this would be enabled both with the Trainer and Fabric.

A contribution like this is very difficult and has a lot of things to consider.

**Comment by rishabbala** [comment]
I think I have sufficient resources to test larger models, and I agree more tests are requried. Let me know if this feature would still be required, if not I can close this PR

**Comment by rishabbala** [comment]
Makes sense, I'll go ahead and close this PR then
